intel-sgx/async-usercalls/src/utils.rs

```diff
@@ -0,0 +1,38 @@
+use std::ops::{Deref, DerefMut};
+use std::os::fortanix_sgx::usercalls::alloc::User;
+use std::os::fortanix_sgx::usercalls::raw::ByteBuffer;
+
+pub(crate) trait MakeSendMarker {}
+
+pub(crate) struct MakeSend<T: MakeSendMarker>(T);
+
+impl<T: MakeSendMarker> MakeSend<T> {
+    pub fn new(t: T) -> Self {
+        Self(t)
+    }
+
+    #[allow(unused)]
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+impl<T: MakeSendMarker> Deref for MakeSend<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T: MakeSendMarker> DerefMut for MakeSend<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+unsafe impl<T: MakeSendMarker> Send for MakeSend<T> {}
+
+impl MakeSendMarker for ByteBuffer {}
+impl MakeSendMarker for User<ByteBuffer> {}
+impl MakeSendMarker for User<[u8]> {}
```

Comment:
What do you mean? I think the existence of this impl is used here: https://github.com/vn971/rust-sgx/blob/vas/async-usercalls-scratch/intel-sgx/async-usercalls/src/provider_api.rs#L27
Or maybe you mean something else?

Reply:
TODO: create an issue. Fixing this requires upstreaming.

Reply:
Documented here: https://github.com/fortanix/rust-sgx/issues/530

Reply:
The description in ticket #530 doesn't make sense. `Send` does not provide any methods. It is define automatically when the compiler considers it appropriate. Unfortunately, it fails here as it contains a pointer to userspace. Pointers typically aren't send because you shouldn't send it over thread boundaries, and thus potentially access them at the same time by these different threads. In this case, it's more complicated. As it's userspace, an attacker could _always_ access it concurrently and its contents should never be trusted. We could just make `User` and `ByteBuffer` implement `Send` instead (but that requires upstream changes).

Reply:
Once [this PR](https://github.com/rust-lang/rust/pull/118241) is upstreamed, these changes can be removed.